File: scientific_computing/L2/ex2.py
import math
import numpy as np
import matplotlib.pyplot as plt

def f(x):
    """
    Defines the function f(x) = exp(x) * log(1 + exp(-x)).
    """
    try:
        return math.exp(x) * math.log(1.0 + math.exp(-x))
    except (ValueError, OverflowError):
        return float('nan')


xs = np.arange(0.0, 40.01, 0.1)
ys = [f(x) for x in xs]

print("Checking for underflow (f(x) = 0.0)...")
for x in np.arange(30.0, 40.01, 0.1):
    val = f(x)
    if val == 0.0:
        print(f"f(x) staje się 0.0 dla x ~= {x:.1f}")
        break

print("\nChecking for loss of stability (deviation from 1.0)...")
for x in np.arange(20.0, 40.01, 0.1):
    val = f(x)
    if abs(val - 1) > 0.01:
        print(f"f(x) traci stabilność dla x ~= {x:.1f}")
        break

fig, ax = plt.subplots(figsize=(10, 6))

ax.plot(xs, ys, label="f(x) = exp(x) * log(1 + exp(-x))")

ax.axhline(1.0, 
           label="Granica analityczna y = 1", 
           linestyle="--",
           color="red")

ax.set_xlabel("x")
ax.set_ylabel("f(x)")
ax.set_title("Zad. 2: Błąd anulowania (Float64)")
ax.legend(loc="lower left")

ax.grid(True, linestyle=':', alpha=0.7)

fig.savefig("zad2_plot.png")

File: scientific_computing/L2/ex2.jl
# Krzysztof Zając

using Plots

f(x) = exp(x) * log(1.0 + exp(-x))

xs = 0.0:0.1:40.0
ys = f.(xs)

for x in 30.0:0.1:40.0
    val = f(x)
    if val == 0.0
        println("f(x) staje się 0.0 dla x ~= $x")
        break
    end
end
for x in 20.0:0.1:40.0
    val = f(x)
    if abs(val - 1) > 0.01
        println("f(x) traci stabilność dla x ~= $x")
        break
    end
end

p = plot(xs, ys, 
    label="f(x) = exp(x) * log(1 + exp(-x))", 
    xlabel="x", 
    ylabel="f(x)",
    title="Zad. 2: Błąd anulowania (Float64)",
    legend=:bottomleft
)
hline!([1.0], label="Granica analityczna y = 1", linestyle=:dash, color=:red)

savefig(p, "ex2_plot_jl.png")